File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/EventHandle/EventDispatch.cpp
﻿#include "MyProject.h"
#include "EventDispatch.h"
#include "Prequisites.h"
#include "EventDispatchFunctionObject.h"
#include "MClassFactory.h"
#include "SafePointer.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(EventDispatch, DelayPriorityHandleMgrBase)

EventDispatch::EventDispatch(int eventId_)
{
	this->mEventId = eventId_;
}

EventDispatch::~EventDispatch()
{
	
}

int EventDispatch::getUniqueId()
{
	return this->mUniqueId;
}

void EventDispatch::setUniqueId(int value)
{
	this->mUniqueId = value;
	this->mHandleList.setUniqueId(mUniqueId);
}

// 相同的函数只能增加一次
void EventDispatch::addEventHandle(EventDispatchDelegate handle)
{
	EventDispatchFunctionObject* funcObject = MY_NEW EventDispatchFunctionObject();
	funcObject->mHandle = handle;

	if (nullptr != handle)
	{
		this->_addObject(funcObject);
	}
	else
	{
		GLogSys->log("Event Handle is nullptr");
	}
}

void EventDispatch::_addObject(IDelayHandleItem* delayObject, float priority)
{
	if (this->_isInDepth())
	{
		Super::_addObject(delayObject, priority);
	}
	else
	{
		// 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
		this->mHandleList.add((EventDispatchFunctionObject*)delayObject);
	}
}

void EventDispatch::removeEventHandle(EventDispatchDelegate handle)
{
	int idx = 0;

	for (idx = 0; idx < mHandleList.count(); ++idx)
	{
		if (this->mHandleList[idx]->mHandle, handle)
		{
			break;
		}
	}
	if (idx < this->mHandleList.count())
	{
		this->_removeObject(this->mHandleList[idx]);
	}
	else
	{
		GLogSys->log("Event Handle not exist");
	}
}

void EventDispatch::_removeObject(IDelayHandleItem* delayObject)
{
	if (this->_isInDepth())
	{
		Super::_removeObject(delayObject);
	}
	else
	{
		if (!this->mHandleList.remove((EventDispatchFunctionObject*)delayObject))
		{
			GLogSys->log("Event Handle not exist");
		}

		MY_SAFE_DISPOSE((EventDispatchFunctionObject*)delayObject);
	}
}

void EventDispatch::dispatchEvent(IDispatchObject* dispatchObject)
{
	//try
	//{
	this->_incDepth();

	for(auto handle : this->mHandleList.getList())
	{
		if (!(GObject*)handle->isClientDispose())
		{
			handle->mHandle(dispatchObject);
		}
	}

	//if (m_luaCSBridgeDispatch != nullptr)
	//{
	//	m_luaCSBridgeDispatch.handleGlobalEvent(mEventId, dispatchObject);
	//}

	this->_decDepth();
	//}
	//catch (Exception ex)
	//{
	//    Ctx.m_instance.m_logSys.catchLog(ex.ToString());
	//}
}

void EventDispatch::clearEventHandle()
{
	if (this->_isInDepth())
	{
		for(auto item : mHandleList.getList())
		{
			this->_removeObject(item);
		}
	}
	else
	{
		this->mHandleList.clear();
	}
}

// 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
bool EventDispatch::isExistEventHandle(EventDispatchDelegate handle)
{
	bool bFinded = false;

	for(auto item : this->mHandleList.getList())
	{
		if (item->mHandle == handle)
		{
			bFinded = true;
			break;
		}
	}

	return bFinded;
}

void EventDispatch::copyFrom(EventDispatch& rhv)
{
	for(auto handle : rhv.getHandleList().getList())
	{
		this->mHandleList.add(handle);
	}
}

MList<EventDispatchFunctionObject*>& EventDispatch::getHandleList()
{
	return this->mHandleList;
}

bool EventDispatch::hasEventHandle()
{
	return this->mHandleList.count() > 0;
}

int EventDispatch::getEventHandleCount()
{
	return this->mHandleList.count();
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/EventHandle/EventDispatch.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.DataStruct.MList");
MLoader("MyLua.Libs.DelayHandle.DelayPriorityHandleMgrBase");
MLoader("MyLua.Libs.EventHandle.EventDispatchFunctionObject");

--[[
    @brief 事件分发器
]]

local M = GlobalNS.Class(GlobalNS.DelayPriorityHandleMgrBase);
M.clsName = "EventDispatch";
GlobalNS[M.clsName] = M;

function M:ctor(eventId_)
    self.mEventId = eventId_;
    self.mHandleList = GlobalNS.new(GlobalNS.MList);
    self.mUniqueId = 0;       -- 唯一 Id ，调试使用
end

function M:dtor()

end

function M:getHandleList()
    return self.mHandleList;
end

function M:getUniqueId()
    return self.mUniqueId;
end

function M:setUniqueId(value)
    self.mUniqueId = value;
    --self.mHandleList.uniqueId = mUniqueId;
end

function M:addEventHandle(pThis, handle, eventId)
    if (nil ~= handle) then
        local funcObject = GlobalNS.new(GlobalNS.EventDispatchFunctionObject);
        funcObject:setFuncObject(pThis, handle, eventId);
        self:addObject(funcObject);
    else
        -- 日志
    end
end

function M:addObject(delayObject, priority)
    if (self:isInDepth()) then
        M.super.addObject(self, delayObject, priority); -- super 使用需要自己填充 Self 参数
    else
        -- 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
        self.mHandleList:add(delayObject);
    end
end

function M:removeEventHandle(handle, pThis, eventId)
    local idx = 0;
	
    for idx = 0, self.mHandleList:count() - 1, 1 do
        if (self.mHandleList:at(idx):isEqual(handle, pThis, eventId)) then
            break;
        end
    end
	
    if (idx < self.mHandleList:count()) then
        self:removeObject(self.mHandleList[idx]);
    else
        -- 日志
    end
end

function M:removeObject(delayObject)
    if (self:isInDepth()) then
        M.super.removeObject(self, delayObject);
    else
        if (self.mHandleList:remove(delayObject) == false) then
            -- 日志
        end
    end
end

function M:dispatchEvent(dispatchObject)
    self:incDepth();

    for _, handle in ipairs(self.mHandleList:list()) do
        if (handle.mIsClientDispose == false) then
            handle:call(dispatchObject);
        end
    end

    self:decDepth();
end

function M:clearEventHandle()
    if (self:isInDepth()) then
        for _, item in ipairs(self.mHandleList:list()) do
            self:removeObject(item);
        end
    else
        self.mHandleList:clear();
    end
end

-- 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
function M:isExistEventHandle(pThis, handle, eventId)
    local isFinded = false;
	
    for _, item in ipairs(self.mHandleList:list()) do
        if (item:isEqual(pThis, handle, eventId)) then
            isFinded = true;
            break;
        end
    end

    return isFinded;
end

function M:copyFrom(rhv)
    for _, handle in ipairs(rhv.handleList:list()) do
        self.mHandleList:add(handle);
    end
end

return M;